### PR TITLE
[MeshModel] Add Schema for Component and Relationship

### LINKED ISSUE #6829: Publish Meshery Model Schemas to JSON Schema Store
#### Current Behavior
Meshmodel schemas are not published.

#### Desired Behavior
Make MeshModel schemas available via the https://github.com/schemastore/schemastore/

#### Acceptance Tests
1. Meshery construct definitions can be validate in IDEs, specifically VS Code.

---
#### Contributor [Guides](https://docs.meshery.io/project/contributing) and [Handbook](https://layer5.io/community/handbook)
- 🛠 [Meshery Build & Release Strategy](https://docs.meshery.io/project/build-and-release)
- 📚 [Instructions for contributing to documentation](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#documentation-contribution-flow)
   - Meshery documentation [site](https://docs.meshery.io/) and [source](https://github.com/meshery/meshery/tree/master/docs)
- 🎨 Wireframes and designs for Meshery UI in [Figma](https://www.figma.com/file/SMP3zxOjZztdOLtgN4dS2W/Meshery-UI)
- 🙋🏾🙋🏼 Questions: [Discussion Forum](https://discuss.layer5.io) and [Community Slack](http://slack.layer5.io)

### PR DESCRIPTION
**Pull request by humblenginr**
This PR fixes #6829 

Some logic in the Component generator should be changed to make the Components adhere to the spec defined in this PR:
1. `display-name` should be changed to `displayName`

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by leecalcote** [comment]
@sudo-NithishKarthik, this is great. We seem to have lost the `pattern` that was defined as valid strings for inputs like displayName.

**Comment by sudo-NithishKarthik** [comment]
> @sudo-NithishKarthik, this is great. We seem to have lost the pattern that was defined as valid strings for inputs like displayName.

It was defined as part of the `definitions` property and was not referenced anywhere, so I thought that was redundant.

**Comment by sudo-NithishKarthik** [comment]
Shall I use the same pattern for all the string inputs in the Schema?

**Comment by leecalcote** [comment]
> Shall I use the same pattern for all the string inputs in the Schema?

@sudo-NithishKarthik, I *think* that doing so only helps bolster the structure of our data and the integrity of our data. I was looking through popular json schemas like those for Prometheus, Argo Events, and GitHub Workflows on the json schemas site/registry and was attempting to glean what best practices (or what I considered to be good practices) I could. I’m hoping that you might peruse some of the better written ones and use them as a model.

**Comment by leecalcote** [comment]
Some of them seems to make Kubernetes resource schemas look weakly (lazily) defined.

**Comment by leecalcote** [comment]
We're in need of a "type" and/or "category" attribute.

**Comment by leecalcote** [comment]
@harkiratsm your input here would be great.

**Comment by harkiratsm** [comment]
Are these jsonSchema compatible with rjsf ? 
how is it like because at this stage I may be misinterpreting. 

If we are using integer make sure use min/max, with string where ever it is possible add ```pattern``` .

**Comment by sudo-NithishKarthik** [comment]
RJSF is compatible with JSONSchema draft 7 which is what we are using to define our schemas and so It can be used with RJSF.

**Comment by leecalcote** [comment]
@harkiratsm you're not able to import these schemas in the rjsf playground? When I remove the schema.meshery.io reference, I'm able to load the component.json schema into the playground.

File: server/meshmodel/schemas/component.json [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      "type": "object",
+      "required": [
+        "name",
+        "version"
+      ],
+      "properties": {
+        "description": {
+          "type": "string"
+        },
+        "name": {
```
Comment (harkiratsm) [inline server/meshmodel/schemas/component.json:30]: Could you add a regex ```pattern``` where ever it possible .

Comment (leecalcote) [reply server/meshmodel/schemas/component.json:30]: @harkiratsm i had one defined in the first commit. You can grab it and update this PR with your additions of pattern regexes. 

What other suggestions do you have? I already suggested this change.

**Comment by leecalcote** [comment]
@sudo-NithishKarthik, @Revolyssup, @Abhishek-kumar09 we have agreed that when you refer specifically to interacting with an API resoruce/property that we will use UpperCamelCase, also known as Pascal case, right?

**Comment by sudo-NithishKarthik** [comment]
Yes, I think.

File: server/meshmodel/schemas/component.json (hunk 1 above)
Comment (harkiratsm) [reply server/meshmodel/schemas/component.json:30]: Not much !

**Review by leecalcote — CHANGES_REQUESTED** [review]
@sudo-NithishKarthik add patterns for string validation and min/max limits for numeric values...

File: server/meshmodel/schemas/component.json [hunk 2]
```diff
@@ -0,0 +1,57 @@
+{
+  "$id": "https://schema.meshery.io/component.json",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "description": "Components are the atomic units for designing infrastructure.",
+  "required": [
+    "apiVersion",
```
[the remark this answers is not in the document]
Comment (sudo-NithishKarthik) [reply server/meshmodel/schemas/component.json:6]: I did not read that comment properly.

**Comment by leecalcote** [comment]
> No, camelCase is what we will be using

Ok.

File: server/meshmodel/schemas/core.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
+{
+  "$id": "https://schema.meshery.io/core.json",
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "description": "Reusable core schema elements",
+  "definitions": {
+    "inputString": {
+      "type": "string",
+      "pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"
```
Comment (leecalcote) [inline server/meshmodel/schemas/core.json:8]: @sudo-NithishKarthik good. Is this pattern not applicable to other strings?

**Review by leecalcote — CHANGES_REQUESTED** [review]

**Comment by leecalcote** [comment]
Please include the `isnamespaced` property in MeshModel modal schema, @Revolyssup

File: server/meshmodel/schemas/core.json (hunk 3 above)
Comment (harkiratsm) [reply server/meshmodel/schemas/core.json:8]: I just made this pattern applicable to the ```name``` property as well.

Comment (leecalcote) [reply server/meshmodel/schemas/core.json:8]: @harkiratsm every single field of type string should have a pattern assigned.

**Comment by harkiratsm** [comment]
new Changes - 
1. Added Description to fields.
2. Added apiVersion & version pattern , defined it under the definitions so it can be reusable.
3. Added name pattern to most of strings, where it make sense.

File: server/meshmodel/schemas/relationship.json [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      "properties": {
+        "from": {
+          "type": "array",
+          "items": {
+            "type": "object",
+            "additionalProperties": false,
+            "properties": {
+              "kind": {
+                "type": "string"
+              },
```
Comment (leecalcote) [inline server/meshmodel/schemas/relationship.json:53]: @harkiratsm should we have ```"pattern": "^[a-zA-Z_][a-zA-Z0-9_]*$"``` here, too?

**Comment by leecalcote** [comment]
Merging for now, understanding that MeshModel will likely move to OpenAPI v3.1.

**Review by leecalcote — APPROVED** [review]

**Pull request merged**